src/librustc_mir/borrow_check/nll/region_infer/error_reporting/outlives_suggestion.rs

```diff
@@ -0,0 +1,428 @@
+            new_name_counter: 0,
+        }
+    }
+
+    /// Add the outlives constraint `fr: outlived_fr` to the set of constraints we need to suggest.
+    pub fn collect_constraint(
+        &mut self,
+        fr: RegionVid,
+        outlived_fr: RegionVid,
+        rinfcx: &RegionInferenceContext<'tcx>,
+        tcx: TyCtxt<'_, '_, 'tcx>,
+    ) {
+        debug!("Collected {:?}: {:?}", fr, outlived_fr);
+
+        self.lookup_or_create_region_name(fr, rinfcx, tcx);
+        self.lookup_or_create_region_name(outlived_fr, rinfcx, tcx);
+
+        if let Some(ref mut outlived_frs) = self.constraints_to_add.get_mut(&fr) {
+            outlived_frs.push(outlived_fr);
+        } else {
+            self.constraints_to_add.insert(fr, vec![outlived_fr]);
+        }
+    }
+
+    /// Either get the existing user-define name of the given region or create one.
+    fn lookup_or_create_region_name(
+        &mut self,
+        fr: RegionVid,
+        rinfcx: &RegionInferenceContext<'tcx>,
+        tcx: TyCtxt<'_, '_, 'tcx>,
+    ) {
+        /// All generated names will be of the form `'PREFIXN`, where `N` is a unique integer.
+        const PREFIX: &str = "generatedl";
+
+        // Already added a name
+        if self.names_to_add.get(&fr).is_some() {
+            return;
+        }
+
+        let has_name = rinfcx
+            .give_name_from_error_region(tcx, self.mir_def_id, fr, &mut 0)
+            .is_some();
+
+        if !has_name {
+            let new_name = format!("'{}{}", PREFIX, self.new_name_counter);
+            debug!("Generate name {} for region {:?}", new_name, fr);
+            self.names_to_add.insert(fr, new_name);
+            self.new_name_counter += 1;
+        }
+    }
+
+    /// If there is a suggestion to emit, add a diagnostic to the buffer.
+    pub fn add_suggestion<'tcx>(
+        &self,
+        rinfcx: &RegionInferenceContext<'tcx>,
+        infcx: &InferCtxt<'_, '_, 'tcx>,
+        errors_buffer: &mut Vec<Diagnostic>,
+    ) {
+        // Create a new diagnostic.
+        let mut diag = DiagnosticBuilder::new(
+            infcx.tcx.sess.diagnostic(),
+            Level::Help,
+            "the following suggestions may resolve your lifetime errors",
+        );
+
+        // Build the diagnostic
+        let mut added_names = self.suggestion_name_implicit_lifetimes(rinfcx, infcx, &mut diag);
+        let mut added_constraints = self.suggestion_outlives_constraints(rinfcx, infcx, &mut diag);
+
+        // Buffer the diagnostic or cancel it if there was no suggestion.
+        if added_names || added_constraints {
+            diag.buffer(errors_buffer);
+        } else {
+            diag.cancel();
+        }
+    }
+
+    /// Add a suggestion to give names to any implicit lifetimes. This may require adding generics
+    /// (i.e. `<'a>`) if all of the lifetimes are implicit. It also requires making lifetimes
+    /// explicit in names.
+    fn suggestion_name_implicit_lifetimes<'tcx>(
```

Comment:
@estebank This function and the next one generally feel really hacky. Do you know if there is a better way to do this?